### PR TITLE
Stg/keycloak unauthorized fix

### PR DESCRIPTION
**Pull request by stompGT**
## Proposed Changes

- Refresh Keycloak ServiceAccount access_token before use.
- Filter requests on API requests to use Keycloak's Devise strategy ( Authorization through `Authorization` header ) 

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Check if route, query, or mutation authorization looks correct.
  - Add tests for authorization, if required.
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Update developer and product docs, where applicable.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Check if new tables or columns that have been added need to be handled in the following services:
  - `Users::DeleteAccountService`
  - `Courses::CloneService`
  - `Courses::DeleteService`
  - `Courses::DemoContentService`
  - `Schools::DeleteService`
- [ ] Check if changes in _packaged_ components have been published to `npm`.
- [ ] Add development seeds for new tables.

### CODE REVIEW

File: lib/devise/strategies/keycloak.rb [hunk 1]
```diff
@@ -3,11 +3,18 @@
 module Devise
   module Strategies
     class Keycloak < Authenticatable
+      def valid?
+        request.path.match?(/^\/api/)
+      end
+
       def authenticate!
+        puts "Authenticate!"
```
Comment (mpraglowski) [inline lib/devise/strategies/keycloak.rb]: No puts/console.log/etc in code please ;)

File: lib/devise/strategies/keycloak.rb [hunk 2]
```diff
@@ -3,11 +3,18 @@
...
+        Rails.logger.debug("Auth header #{request.headers['Authorization']}")
```
Comment (mpraglowski) [inline lib/devise/strategies/keycloak.rb]: Do we need this noise in logs?

File: app/controllers/api/schools/courses_controller.rb [hunk 3]
```diff
@@ -18,20 +20,26 @@ def students
       end
 
       def create_students
-        form = ::Schools::Founders::CreateForm.new(Reform::OpenForm.new)
+        form = Students::CreateForm.new(Reform::OpenForm.new)
 
-        response = if form.validate(params)
+        Rails.logger.debug("Students params: #{students_params}")
```
Comment (mpraglowski) [inline app/controllers/api/schools/courses_controller.rb]: I see no point for this.

File: app/controllers/api/schools/courses_controller.rb [hunk 4]
```diff
@@ -1,12 +1,14 @@
 module Api
   module Schools
-    class CoursesController < SchoolsController
+    class CoursesController < ApplicationController
       skip_before_action :verify_authenticity_token
+
```
Comment (mpraglowski) [inline app/controllers/api/schools/courses_controller.rb]: the same as in #53  ?

File: app/services/users/validate_reset_token_service.rb [hunk 5]
```diff
@@ -22,10 +22,31 @@ def user
     def valid_request?
       return false if user.blank?
 
-      return false if user.reset_password_sent_at.blank?
+      return false if user.reset_password_sent_at.blank? && time_limitation?
```
Comment (mpraglowski) [inline app/services/users/validate_reset_token_service.rb]: maybe:

```ruby
return true unless time_limitation?
return false if user.reset_password_sent_at.blank?
valid_time?
 ```

File: app/services/users/validate_reset_token_service.rb [hunk 6]
```diff
@@ -22,10 +22,31 @@ def user
...
+    def time_limitation?
+      ENV.fetch('RESET_PASSWORD_TOKEN_TIME_LIMIT') { '' }.present?
+    end
+
+    def time_limit_minutes
+      env_var = ENV.fetch('RESET_PASSWORD_TOKEN_TIME_LIMIT') { '' } 
+      if env_var.present?
+        env_var.to_i.minutes
+      else
+        0.minutes
```
Comment (mpraglowski) [inline app/services/users/validate_reset_token_service.rb]: why if we have the `time_limitation?` check ? this should not be possible

File: config/environments/production.rb [hunk 7]
```diff
@@ -150,7 +150,7 @@
   require_relative '../../lib/rack_throttle/rules'
 
   rules = [
-    { method: "POST", limit: ENV['GRAPH_API_RATE_LIMIT'] },
+    { method: "POST", limit: ENV['GRAPH_API_RATE_LIMIT'] || 1 },
```
Comment (mpraglowski) [inline config/environments/production.rb:153]: Why in this PR? How this is connected with Keycloak ?

Comment (stompGT) [reply config/environments/production.rb:153]: Not having a default value for that limit was preventing the creation of the students through the API. It was causing any `POST` call to the API to be denied with a `503 Rate limit exceeded` . Is this a bad change? Should we rely on just the value of the env var and let `POSTs` be blocked if it's not defined?

File: app/services/users/validate_reset_token_service.rb (hunk 5 above)
Comment (stompGT) [reply app/services/users/validate_reset_token_service.rb]: Removed when rebased from master.

File: app/services/users/validate_reset_token_service.rb (hunk 6 above)
Comment (stompGT) [reply app/services/users/validate_reset_token_service.rb]: Removed when rebased from master.

File: app/controllers/api/schools/courses_controller.rb (hunk 3 above)
Comment (stompGT) [reply app/controllers/api/schools/courses_controller.rb]: Removed

File: lib/devise/strategies/keycloak.rb (hunk 2 above)
Comment (stompGT) [reply lib/devise/strategies/keycloak.rb]: Removed

File: lib/devise/strategies/keycloak.rb (hunk 1 above)
Comment (stompGT) [reply lib/devise/strategies/keycloak.rb]: Removed

File: config/environments/production.rb (hunk 7 above)
Comment (mpraglowski) [reply config/environments/production.rb:153]: No. I was just curious how this affects keycloak integration. It's good to have defaults not crash whole app like if some VAPID keys are not defined ;)

File: lib/keycloak.rb [hunk 8]
```diff
@@ -161,7 +196,8 @@ def user_info(access_token)
       if res.status == 200
         MultiJson.load(res.body)
       else
-        raise FailedRequestError.new 'Failed to set user password'
+        Rails.logger.debug(res.body)
```
Comment (mpraglowski) [inline lib/keycloak.rb]: Can we get rid of all logs like this? Or we rely on them in production?

File: lib/keycloak.rb [hunk 9]
```diff
@@ -269,7 +305,7 @@ def user_sign_out(refresh_token)
 
     private
     def user_by_token(token)
-      @users.find{|_, user| user[:token] == token}
+      @users.find{|_, user| user[:token] == token}[1]
```
Comment (mpraglowski) [inline lib/keycloak.rb]: I still don't get that code. @users is a hash of hashes, what you expect to get from `{username: email, email: email, ... }[1]` ?

Comment (mpraglowski) [reply lib/keycloak.rb]: Oh stupid me - the result of find here is `[key,value]` :)

**Pull request merged**